node/core/dispute-coordinator/src/real/mod.rs

```diff
@@ -499,10 +502,212 @@ async fn handle_new_activations(
+	state: &mut State,
+	new_leaf: Hash,
+	now: u64,
+	metrics: &Metrics,
+) -> Result<(), Error> {
+	// obtain the concluded disputes as well as the candidate backing votes
+	// from the new leaf
+	let ScrapedOnChainVotes { session, backing_validators, disputes } = {
+		let (tx, rx) = oneshot::channel();
+		ctx.send_message(RuntimeApiMessage::Request(
+			new_leaf,
+			RuntimeApiRequest::FetchOnChainVotes(tx),
+		))
+		.await;
+		match rx.await {
+			Ok(Ok(Some(val))) => val,
+			Ok(Ok(None)) => {
+				tracing::trace!(
+					target: LOG_TARGET,
+					relay_parent = ?new_leaf,
+					"No on chain votes stored for relay chain leaf");
+				return Ok(())
+			},
+			Ok(Err(e)) => {
+				tracing::debug!(
+					target: LOG_TARGET,
+					relay_parent = ?new_leaf,
+					error = ?e,
+					"Could not retrieve on chain votes due to an API error");
+				return Ok(())
+			},
+			Err(e) => {
+				tracing::debug!(
+					target: LOG_TARGET,
+					relay_parent = ?new_leaf,
+					error = ?e,
+					"Could not retrieve onchain votes due to oneshot cancellation");
+				return Ok(())
+			},
+		}
+	};
+
+	if backing_validators.is_empty() && disputes.is_empty() {
+		return Ok(())
+	}
+
+	// Obtain the session info, for sake of `ValidatorId`s
+	// either from the rolling session window.
+	// Must be called _after_ `fn cache_session_info_for_head`
+	// which guarantees that the session info is available
+	// for the current session.
+	let session_info: SessionInfo =
+		if let Some(session_info) = state.rolling_session_window.session_info(session) {
+			session_info.clone()
+		} else {
+			tracing::warn!(
+				target: LOG_TARGET,
+				relay_parent = ?new_leaf,
+				"Could not retrieve session info from rolling session window");
+			return Ok(())
+		};
+
+	// Scraped on-chain backing votes for the candidates with
+	// the new active leaf as if we received them via gossip.
+	for (candidate_receipt, backers) in backing_validators {
+		let candidate_hash = candidate_receipt.hash();
+		let statements =
+			backers.into_iter().filter_map(|(validator_index, validator_signature)| {
+				let validator_public: ValidatorId = session_info
+					.validators
+					.get(validator_index.0 as usize)
+					.or_else(|| {
+						tracing::error!(
+						target: LOG_TARGET,
+						relay_parent = ?new_leaf,
+						"Missing public key for validator {:?}",
+						&validator_index);
+						None
+					})
+					.cloned()?;
+				let signed_dispute_statement =
+					SignedDisputeStatement::new_unchecked_from_trusted_source(
+						// TODO verify this is the correct statement
+						DisputeStatement::Valid(ValidDisputeStatementKind::BackingValid(new_leaf)),
+						candidate_hash,
+						session,
+						validator_public,
+						validator_signature,
+					);
+				Some((signed_dispute_statement, validator_index))
+			});
+		let import_result = handle_import_statements(
+			ctx,
+			overlay_db,
+			state,
+			candidate_hash,
+			MaybeCandidateReceipt::AssumeBackingVotePresent,
```

Comment:
Isn't the candidate receipt unconditionally present at this point? It's part of the `backing_validators` object that was iterated over.

Reply:
The only point I can think of when the candidate receipt wouldn't be present is if we're scraping a remote dispute from the chain and haven't seen any of:
1. a fork where the block was backed
2. an off-chain backing message
3. the dispute messages off-chain